Added JSON serialization and multi-recipient support to JWE

spivachuk opened the pull request:
**What kind of change does this PR introduce?** (check at least one)

- [ ] Bugfix
- [x] Feature
    - Added JSON serialization and multi-recipient support to JWE
- [ ] Code style update
- [ ] Refactor
- [ ] Other, please describe:

**Does this PR introduce a breaking change?** (check one)

- [ ] Yes
- [x] No

---

- [x] You consent that the copyright of your pull request source code belongs to Authlib's author.

spivachuk committed:
Added JSON serialization and multi-recipient support to JWE

ashcherbakov commented on authlib/jose/drafts/_jwe_algorithms.py:
```diff
@@ -32,6 +32,15 @@ def prepare_key(self, raw_data):
         return ECKey.import_key(raw_data)
 
+    def generate_preset(self, enc_alg, key):
+        epk = key.generate_key(key['crv'], is_private=True)
```
Should we call `self._generate_ephemeral_key(key)` instead?

spivachuk replied:
Fixed.

ashcherbakov commented on authlib/jose/drafts/_jwe_algorithms.py:
```diff
@@ -88,19 +97,27 @@ def deliver_at_recipient(self, recipient_key, sender_static_pubkey, sender_ephem
-        h = self._prepare_headers(sender_key, epk)
+        if preset and 'epk' in preset:
+            epk = preset['epk']
+            h = {}
```
Should we get the header from `preset`?

spivachuk replied:
No, because here we construct specific recipient's unprotected header.

ashcherbakov commented on authlib/jose/rfc7516/jwe.py:
```diff
@@ -139,6 +141,145 @@ def serialize_compact(self, protected, payload, key, sender_key=None):
+        recipients = header_obj.get('recipients')
+        if recipients is None:
+            recipients = [{} for _ in keys]
+        for i in range(len(recipients)):
```
```suggestion
        for recipient in recipients:
```

spivachuk replied:
No, otherwise replacement of a recipient being None with an empty dict will take no effect.

ashcherbakov commented on authlib/jose/rfc7516/jwe.py:
```diff
@@ -139,6 +141,145 @@ def serialize_compact(self, protected, payload, key, sender_key=None):
+        enc = self.get_header_enc(shared_header)
+        zip_alg = self.get_header_zip(shared_header)
+
+        self._validate_sender_key(sender_key, alg)
```
Do we validate recipient keys and check that sender and recipient keys are on the same curve?

spivachuk replied:
This is done by the underlying crypto library.

ashcherbakov commented on authlib/jose/rfc7516/jwe.py:
```diff
@@ -197,6 +338,110 @@ def deserialize_compact(self, s, key, decode=None, sender_key=None):
             payload = decode(payload)
         return {'header': protected, 'payload': payload}
 
+    def deserialize_json(self, obj, key, decode=None, sender_key=None):
```
In many cases we will need to parse the message first, get the recipient key IDs, and only then call descrypt/deserialize for the found recipient key values. So, we will need a separate method to parse the message.

spivachuk replied:
Added `parse_json` method.

ashcherbakov commented on authlib/jose/rfc7516/jwe.py:
```diff
@@ -197,6 +338,110 @@ def deserialize_compact(self, s, key, decode=None, sender_key=None):
+        enc = self.get_header_enc(shared_header)
+        zip_alg = self.get_header_zip(shared_header)
+
+        self._validate_sender_key(sender_key, alg)
```
Do we validate recipient keys and check that sender and recipient keys are on the same curve?

spivachuk replied:
This is done by the underlying crypto library.

ashcherbakov commented on authlib/jose/rfc7516/jwe.py:
```diff
@@ -197,6 +338,110 @@ def deserialize_compact(self, s, key, decode=None, sender_key=None):
+            return alg.unwrap(enc, ek, header, key)
+
+        def _unwrap_for_some_recipient(unwrap_func):
+            for recipient in recipients:
```
I believe we must be able to provide a key with a key ID, so that we can avoid that iteration, and try to decrypt for the given key only.

spivachuk replied:
Added such logic.

ashcherbakov commented on authlib/jose/rfc7518/jwe_algs.py:
```diff
@@ -187,6 +213,15 @@ def prepare_key(self, raw_data):
         return ECKey.import_key(raw_data)
 
+    def generate_preset(self, enc_alg, key):
+        epk = key.generate_key(key['crv'], is_private=True)
```
Should we call self._generate_ephemeral_key(key) instead?

spivachuk replied:
Fixed.

ashcherbakov commented on tests/core/test_jose/test_jwe.py:
```diff
@@ -886,6 +886,73 @@ def test_ecdh_1pu_jwe_with_okp_keys_in_key_agreement_with_key_wrapping_mode(self
                 rv = jwe.deserialize_compact(data, bob_key, sender_key=alice_key)
                 self.assertEqual(rv['payload'], b'hello')
 
+    def test_ecdh_1pu_jwe_json_serialization(self):
```
Please add more tests for JSON serialization and multi-recipient.
Please add tests for the test vectors from ECDH-1PU and ECDH-ES where JSON serialization is used.

ashcherbakov replied:
Also, please add tests with edge cases, where unexpected parameters are passed:
- keys from different curves
- invalid/unexpected/missing values in the input header
- etc.

spivachuk replied:
Done.

spivachuk committed:
JSON serialization and multi-recipient support for JWE

- Made some corrections in serialize_json and deserialize_json methods of JsonWebEncryption class.
- Added some tests for JSON serialization and multi-recipient support for JWE.
- Made some minor corrections.

spivachuk committed:
JSON serialization and multi-recipient support for JWE

- Added more tests.
- Fixed found bugs.
- Added a method for parsing a JSON message to JsonWebEncryption class.

spivachuk committed:
JSON serialization and multi-recipient support for JWE

- Added generic serialize and deserialize methods to JsonWebEncryption class and unit tests for them.
- Added a unit test for JsonWebEncryption.deserialize_json method called with JSON string.

spivachuk committed:
JSON serialization and multi-recipient support for JWE

- Added docstrings to new public methods of JsonWebEncryption class.
- Added more unit tests for multi-recipient support in JWE.

ashcherbakov commented on tests/core/test_jose/test_jwe.py:
```diff
@@ -1139,3 +2197,384 @@ def test_dir_alg_c20p(self):
+        data_as_string = json.dumps(data)
+
+        rv = jwe.deserialize(data_as_string, bob_key, sender_key=alice_key)
+        self.assertEqual(rv['payload'], b'hello')
```
2580 lines...
Did you consider split the test into multiple files?

ashcherbakov commented on tests/core/test_jose/test_jwe.py:
```diff
@@ -405,6 +420,229 @@ def test_ecdh_es_jwe_with_okp_key_in_key_agreement_with_key_wrapping_mode(self):
+        bob_key = OKPKey.import_key({
+            "kty": "OKP",
+            "crv": "X25519",
+            "kid": "bob-key-2",
```
Is the only way to specify a KID for decryption is within a JWK object? If so, it may not work for DIDComm.
See https://www.w3.org/TR/did-spec-registries/#jsonwebkey2020
There JWK `kid` is a relative value, we can not resolve a DID based on it. What we will specify as a `kid` in per -recipient header is verification method's `id`, that is DID URL for which we can resolve the DID and find the corresponding key.

spivachuk replied:
Supported a separately provided recipient kid in deserialize methods of `JsonWebEncryption` class.

ashcherbakov commented on tests/core/test_jose/test_jwe.py:
```diff
@@ -1139,3 +2197,384 @@ def test_dir_alg_c20p(self):
+
+        rv = jwe.deserialize(data, bob_key, sender_key=alice_key)
+        self.assertEqual(rv['payload'], b'hello')
+
```
Please add the following tests:
1) Decrypt test: Some random values in JSON for decryption (not from the spec). Expected: ignored and decrypt successful?
2) Decrypt test: ciphertext is invalid. Expected: exception
3) Decrypt test: no epk in protected header.  Expected: exception
4) Decrypt test: `encrypted_key` value for the the given recipient key is invalid.  Expected: exception
5) Decrypt test: `encrypted_key` value for the the given recipient key is valid, but for another recipient not passed to decrypt method it is not valid  Expected: success

spivachuk replied:
Done except for Item 3 because modifying protected header of JWE message (removing `epk` header field before the message decryption) breaks the message consistency (i.e. consistency with AAD, ciphertext, tag).

spivachuk committed:
JSON serialization and multi-recipient support for JWE

- Supported a separately provided recipient kid in deserialize methods of JsonWebEncryption class.

spivachuk committed:
Corrected finding recipient by key on decryption of JSON-serialized JWE according to the recent changes

spivachuk committed:
JSON serialization and multi-recipient support for JWE

- Corrected unit tests of JWE correspondingly to the recent changes.
- Added prohibition of unknown members in JSON-serialized JWE message and in `recipients` list element.
- Added more unit tests for JWE.

ashcherbakov commented on tests/core/test_jose/test_jwe.py:
```diff
@@ -238,6 +239,205 @@ def test_aes_gcm_jwe_invalid_key(self):
+        protected = {
+            "alg": "ECDH-ES+A128KW",
+            "enc": "A128GCM",
+            "foo": "bar"
```
Do we really need to fail here? Does JOSE spec explicitly restrict any values besides the ones listed there?
BTW In Numbus unknown fields are just ignored, there is no failure.

spivachuk replied:
This behavior is controlled by `private_headers` parameter of `JsonWebEncryption` constructor.

ashcherbakov commented on tests/core/test_jose/test_jwe.py:
```diff
@@ -238,6 +239,205 @@ def test_aes_gcm_jwe_invalid_key(self):
+        )
+
+    def test_serialize_json_fails_if_protected_header_contains_unknown_field_while_private_fields_restricted(self):
+        jwe = JsonWebEncryption(private_headers=[])
```
What does `private_fields_restricted` means here?

spivachuk replied:
The mode of `JsonWebEncryption` class which restricts the set of allowed private header fields. Here the set of allowed private header fields is empty (since `private_headers` argument of `JsonWebEncryption` constructor is empty).

spivachuk committed:
JSON serialization and multi-recipient support for JWE

- Removed prohibition of unknown members in JSON-serialized JWE message and in `recipients` list element.

spivachuk committed:
Added one more test for JWE

spivachuk committed:
Made minor correction in unit tests

spivachuk committed:
Added more tests for JWE
